drivers/platform/x86/thinkpad-wmi.c

```diff
@@ -348,77 +340,58 @@ static int thinkpad_wmi_bios_setting(int item, char **value)
 	if (ACPI_FAILURE(status))
 		return -EIO;
 
-	return thinkpad_wmi_extract_output_string(&output, value);
+	return thinkpad_wmi_check_output(&output, value);
 }
 
-static int thinkpad_wmi_get_bios_selections(const char *item, char **value)
+static int thinkpad_wmi_password_settings(struct thinkpad_wmi_pcfg *pcfg)
 {
-	const struct acpi_buffer input = { strlen(item), (char *)item };
+	const union acpi_object *obj;
+
 	struct acpi_buffer output = { ACPI_ALLOCATE_BUFFER, NULL };
 	acpi_status status;
 
-	status = wmi_evaluate_method(LENOVO_GET_BIOS_SELECTIONS_GUID,
-				     0, 0, &input, &output);
-
+	status = wmi_query_block(LENOVO_BIOS_PASSWORD_SETTINGS_GUID, 0,
+				 &output);
 	if (ACPI_FAILURE(status))
 		return -EIO;
 
-	return thinkpad_wmi_extract_output_string(&output, value);
+	obj = output.pointer;
+	if (!obj || obj->type != ACPI_TYPE_BUFFER || !obj->buffer.pointer)
+		return -EIO;
+	if (obj->buffer.length != sizeof(*pcfg)) {
+		pr_warn("Unknown pcfg buffer length %d\n", obj->buffer.length);
+		kfree(obj);
+		return -EIO;
+	}
+
+	memcpy(pcfg, obj->buffer.pointer, obj->buffer.length);
+	kfree(obj);
+	return 0;
 }
 
 static int thinkpad_wmi_set_bios_settings(const char *settings)
 {
-	return thinkpad_wmi_simple_call(LENOVO_SET_BIOS_SETTINGS_GUID,
-					settings);
+	return thinkpad_wmi_call(LENOVO_SET_BIOS_SETTINGS_GUID, settings, NULL);
```

Comment:
I like thinkpad_wmi_simple_call() better than having to specify NULL all the time

Reply:
done